File: backend/services/user_service.js
const User = require('../models/User');
const bcrypt = require('bcryptjs');
const jwt = require('jsonwebtoken');
const config = require('../config');

// generate jwt token for authenticated user
function generateToken(user) {
  const payload = {
    user_id: user._id.toString(),
    role: user.role,
    exp: Math.floor(Date.now() / 1000) + (24 * 60 * 60) // 24 hours
  };
  return jwt.sign(payload, config.JWT_SECRET_KEY);
}

// register new customer
async function registerCustomer(data) {
  const { email, password, name } = data;
  
  const existingUser = await User.findOne({ email });
  if (existingUser) {
    throw new Error("Email already registered");
  }
  
  const hashed = await bcrypt.hash(password, 10);
  
  const user = new User({
    email,
    password_hash: hashed,
    name,
    role: "Customer",
    status: "PendingApproval"
  });
  
  await user.save();
  
  return { message: "Registration successful, pending approval.", user_id: user._id.toString() };
}

async function processRegistrationApproval(managerId, userId, decision, reason = null) {
  const manager = await User.findById(managerId);
  if (!manager || manager.role !== "Manager") {
    return { error: "Unauthorized. Only managers can approve registrations." }, 403;
  }
  
  const user = await User.findById(userId);
  if (!user) {
    return { error: "User not found." }, 404;
  }
  
  if (decision === "APPROVE") {
    user.status = "Active";
  } else if (decision === "REJECT") {
    user.status = "Rejected";
    user.rejectionReason = reason;
  } else {
    return { error: "Invalid decision. Use 'APPROVE' or 'REJECT'." }, 400;
  }
  
  await user.save();
  return { message: `User ${decision.toLowerCase()}ed.` }, 200;
}

async function loginUser(email, password) {
  const user = await User.findOne({ email });
  if (!user) {
    return { error: "Invalid email or password." }, 401;
  }
  
  if (["Rejected", "Blacklisted"].includes(user.status)) {
    return { error: "Account is not allowed to login." }, 403;
  }
  
  const isValid = await bcrypt.compare(password, user.password_hash);
  if (!isValid) {
    return { error: "Invalid email or password." }, 401;
  }
  
  const token = generateToken(user);
  return { token, role: user.role, user_id: user._id.toString() }, 200;
}

async function depositMoney(userId, amount) {
  const user = await User.findById(userId);
  if (!user) {
    return { error: "User not found." }, 404;
  }
  
  if (amount <= 0) {
    return { error: "Deposit amount must be positive." }, 400;
  }
  
  user.balance += amount;
  await user.save();
  return { message: `Deposited $${amount.toFixed(2)}. New balance: $${user.balance.toFixed(2)}.` }, 200;
}

async function applyWarning(userId, reason = "") {
  const user = await User.findById(userId);
  if (!user) {
    return;
  }
  
  user.warningCount += 1;
  
  if (["Customer", "VIP"].includes(user.role)) {
    if (user.role === "VIP") {
      if (user.warningCount >= 2) {
        user.isVIP = false;
        user.role = "Customer";
        user.warningCount = 0;
      }
    } else {
      if (user.warningCount >= 3) {
        user.status = "Deregistered";
        await user.save();
        return;
      }
    }
  }
  
  await user.save();
}

async function applyComplaintEffect(employeeId, entityType, weight) {
  const employee = await User.findById(employeeId);
  if (!employee) {
    return false;
  }
  
  employee.netComplaints += weight;
  
  if (["Chef", "DeliveryPerson"].includes(entityType) && employee.netComplaints >= 3) {
    if (employee.demotionsCount === 0) {
      employee.role = `Demoted_${entityType}`;
      employee.demotionsCount += 1;
      employee.netComplaints = 0;
    } else if (employee.demotionsCount >= 1) {
      employee.status = "Terminated";
    }
  }
  
  await employee.save();
  return true;
}

async function updateVipStatus(customerId) {
  const user = await User.findById(customerId);
  if (!user || !["Customer", "VIP"].includes(user.role)) {
    return false;
  }
  
  if (user.warningCount > 0) {
    return false;
  }
  
  if (user.totalSpent > 100 || user.orderCount >= 3) {
    user.isVIP = true;
    user.role = "VIP";
    await user.save();
    return true;
  }
  return false;
}

async function blacklistUser(managerId, userId) {
  const manager = await User.findById(managerId);
  if (!manager || manager.role !== "Manager") {
    return { error: "Unauthorized. Only managers can blacklist users." }, 403;
  }
  
  const user = await User.findById(userId);
  if (!user) {
    return { error: "User not found." }, 404;
  }
  
  user.status = "Blacklisted";
  await user.save();
  return { message: "User has been blacklisted." }, 200;
}

module.exports = {
  generateToken,
  registerCustomer,
  processRegistrationApproval,
  loginUser,
  depositMoney,
  applyWarning,
  applyComplaintEffect,
  updateVipStatus,
  blacklistUser
};

File: backend/services/user_service.py
from models.user import User
from werkzeug.security import generate_password_hash, check_password_hash
from config import JWT_SECRET_KEY
import jwt
import datetime

# Generate JWT token for authenticated user
def generate_token(user):
    payload = {
        "user_id": str(user.id),
        "role": user.role,
        "exp": datetime.datetime.utcnow() + datetime.timedelta(hours=24)
    }
    token = jwt.encode(payload, JWT_SECRET_KEY, algorithm='HS256')
    return token

# Register new customer
def register_customer(data):
    email = data.get("email")
    password = data.get("password")
    name = data.get("name")

    if User.objects(email=email).first():
        raise ValueError("Email already registered")
    
    hashed = generate_password_hash(password)

    user = User(
        email=email,
        password_hash=hashed,
        name=name,
        role="Customer",
        status="PendingApproval"
    ).save()

    return {"message": "Registration successful, pending approval.", "user_id": str(user.id)}

def process_registration_approval(manager_id, user_id, decision, reason=None):
    # Verify manager role
    manager = User.objects(id=manager_id).first()
    if not manager or manager.role != "Manager":
        return {"error": "Unauthorized. Only managers can approve registrations."}, 403
    
    user = User.objects(id=user_id).first()
    if not user:
        return {"error": "User not found."}, 404
    
    if decision == "APPROVE":
        user.status = "Active"
    elif decision == "REJECT":
        user.status = "Rejected"
        user.rejectionReason = reason
    else:
        return {"error": "Invalid decision. Use 'APPROVE' or 'REJECT'."}, 400
    
    user.save()
    return {"message": f"User {decision.lower()}ed."}, 200
    
def login_user(email, password):
    user = User.objects(email=email).first()
    if not user:
        return {"error": "Invalid email or password."}, 401
        
    if user.status in ["Rejected", "Blacklisted"]:
        return {"error": "Account is not allowed to login."}, 403
        
    if not check_password_hash(user.password_hash, password):
        return {"error": "Invalid email or password."}, 401
        
    token = generate_token(user)
    return {"token": token, "role": user.role, "user_id": str(user.id)}, 200
    
def deposit_money(user_id, amount):
    user = User.objects(id=user_id).first()
    if not user:
        return {"error": "User not found."}, 404
        
    if amount <= 0:
        return {"error": "Deposit amount must be positive."}, 400
        
    user.balance += amount
    user.save()
    return {"message": f"Deposited ${amount:.2f}. New balance: ${user.balance:.2f}."}, 200

def apply_warning(user_id, reason=""):
    """
    UC-01 A3: Warnings
    - Regular Customers: 3 warnings → Deregistered
    - VIPs: 2 warnings → Demoted to Customer (warnings cleared)
    """
    user = User.objects(id=user_id).first()
    if not user:
        return
        
    user.warningCount += 1

    if user.role in ["Customer", "VIP"]:
        if user.role == "VIP": # Warning for VIP customer
            if user.warningCount >= 2:
                user.isVIP = False
                user.role = "Customer"
                user.warningCount = 0
        else: # Regular customer
            if user.warningCount >= 3:
                user.status = "Deregistered"
                user.save()
                return
    
    user.save()

def apply_complaint_effect(employee_id, entity_type, weight):
    """
    Apply complaint effect to employee (Chef or DeliveryPerson)
    Rules:
    - netComplaints += weight
    - If netComplaints >= 3:
      - First time: Demote (role = "Demoted_Chef" or "Demoted_DeliveryPerson")
      - Second time: Terminate (status = "Terminated")
    """
    employee = User.objects(id=employee_id).first()
    if not employee:
        return False
    
    employee.netComplaints += weight
    
    # Only apply demotion/termination to Chef and DeliveryPerson
    if entity_type in ["Chef", "DeliveryPerson"] and employee.netComplaints >= 3:
        if employee.demotionsCount == 0:
            # First demotion
            employee.role = f"Demoted_{entity_type}"
            employee.demotionsCount += 1
            employee.netComplaints = 0  # Reset after demotion
        elif employee.demotionsCount >= 1:
            # Second demotion = termination
            employee.status = "Terminated"
    
    employee.save()
    return True

def update_vip_status(customer_id):
    user = User.objects(id=customer_id).first()
    if not user or user.role not in ["Customer", "VIP"]:
        return False
    
    if user.warningCount > 0:
        return False
    
    if user.totalSpent > 100 or user.orderCount >= 3:
        user.isVIP = True
        user.role = "VIP"
        user.save()
        return True
    return False
    
def blacklist_user(manager_id, user_id):
    manager = User.objects(id=manager_id).first()
    if not manager or manager.role != "Manager":
        return {"error": "Unauthorized. Only managers can blacklist users."}, 403
    
    user = User.objects(id=user_id).first()
    if not user:
        return {"error": "User not found."}, 404
    
    user.status = "Blacklisted"
    user.save()
    return {"message": "User has been blacklisted."}, 200
            
